compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -5785,11 +5785,23 @@ public boolean isNonNilSimpleBasicTypeOrString(BType bType) {
             Set<BType> memberTypes = ((BUnionType) type).getMemberTypes();
             for (BType member : memberTypes) {
                 BType memType = getReferredType(member);
+                if (memType.tag == TypeTags.FINITE || memType.tag == TypeTags.UNION) {
+                    isNonNilSimpleBasicTypeOrString(memType);
+                    continue;
+                }
                 if (memType.tag == TypeTags.NIL || !isSimpleBasicType(memType.tag)) {
                     return false;
                 }
             }
             return true;
+        } else if (type.tag == TypeTags.FINITE) {
+            for (BLangExpression expression: ((BFiniteType) type).getValueSpace()) {
+                BType exprType = getReferredType(expression.getBType());
+                if (exprType.tag == TypeTags.NIL || !isSimpleBasicType(exprType.tag)) {
+                    return false;
```

Comment:
Shall we add a negative test for when the finite type includes nil?

Reply:
@rpjayasekara Can you address the negative tests here? https://github.com/ballerina-platform/ballerina-lang/issues/34163